src/nameresolution/builtin.rs

```diff
@@ -79,30 +90,65 @@ pub fn import_prelude<'a>(resolver: &mut NameResolver, cache: &mut ModuleCache<'
 ///
 /// The builtin string type is defined here as:
 ///
-/// type string = c_string: ref char, length: usz
-///
-/// TODO: The C-string field probably shouldn't be region-allocated with ref (?).
-///       This container type likely needs to change.
+/// type string = c_string: ptr char, length: usz
 fn define_string<'a>(cache: &mut ModuleCache<'a>) -> Type {
     let location = Location::builtin();
 
-    let ref_type = Type::Ref(cache.next_type_variable_id(LetBindingLevel(1)));
+    let ptr_type = Type::Primitive(PrimitiveType::Ptr);
     let char_type = Type::Primitive(PrimitiveType::CharType);
-    let c_string_type = Type::TypeApplication(Box::new(ref_type), vec![char_type]);
+    let c_string_type = Type::TypeApplication(Box::new(ptr_type), vec![char_type]);
 
     let length_type = Type::Primitive(PrimitiveType::IntegerType(IntegerKind::Usz));
 
-    let string = cache.push_type_info("string".into(), vec![], location);
-    assert_eq!(string, STRING_TYPE);
+    let name = "string".to_string();
+    let string_id = cache.push_type_info(name.clone(), vec![], location);
+    assert_eq!(string_id, STRING_TYPE);
+    let string = Type::UserDefinedType(STRING_TYPE);
 
-    cache.type_infos[string.0].body = TypeInfoBody::Struct(vec![
-        Field { name: "c_string".into(), field_type: c_string_type, location },
-        Field { name: "length".into(),   field_type: length_type,   location },
+    let fields = TypeInfoBody::Struct(vec![
+        Field { name: "c_string".into(), field_type: c_string_type.clone(), location },
+        Field { name: "length".into(),   field_type: length_type.clone(),   location },
     ]);
 
-    Type::UserDefinedType(STRING_TYPE)
+    let constructor = cache.push_definition(&name, false, Location::builtin());
+    assert_eq!(constructor, STRING_ID);
+    let constructor_type = Type::Function(vec![c_string_type, length_type], Box::new(string.clone()), false);
+
+    cache.definition_infos[constructor.0].typ = Some(constructor_type);
+    cache.definition_infos[constructor.0].definition = Some(DefinitionKind::TypeConstructor { name, tag: None });
+
+    cache.type_infos[string_id.0].body = fields;
+    string
 }
 
+/// Defining the 'ptr' builtin type
+///
+/// type ptr a
+fn define_ptr<'a>(cache: &mut ModuleCache<'a>) {
```

Comment:
Mentioned this in another comment but just so it shows up in the review I'll repeat it here: lets remove this function